src/shared.ts

```diff
@@ -1,26 +1,46 @@
-export interface State<T> {
-  val: T
-  readonly oldVal: T
-}
+import type { Primitive, Props, PropValue, State, StateView } from './van-plate'
+export type { Primitive, Props, PropValue, State, StateView }
+interface HasFirstChild {firstChild?: unknown}
+
+type NodeType<ElementType extends HasFirstChild> =
+  Omit<ElementType["firstChild"], "after" | "before" | "remove" | "replaceWith">
+
+export type ValidChildDomValue<ElementType = any, TextNodeType = any> =
```

Comment:
Why having the type parameters default to `any`?

Reply:
I'm using the examples I provided to guide development. `van-plate.d.ts` `ValidChildDomValue` is not a generic type while `mini-van.d.ts` `ValidChildDomValue` has two generics. `any` is not a requirement. A placeholder default type could be used instead. There may be a better pattern for the type, but this is what I came up with to solve the type issue in the [examples](https://github.com/vanjs-org/mini-van/pull/8#issuecomment-1788093557).

Reply:
I think `any` is too permissive. I will put more work into this.

Reply:
The typical use case of `shared.ts` is for shared components between SSR and CSR, with the following pattern for example:

```js
const Component(van: VanObj) {
  ...
}
```

For this PR to provide better type checking than status-quo, the default `ElementType` and `TextNodeType` need to have meaningful type information to enable the checking.

I was trying to solve this problem, but found it is indeed very hard to provide a meaningful `ElementType` and `TextNodeType` that are shared between VanJS, `van-plate` and `mini-van`. Thus I gave up on it and chose to enforce just some basic type checking in `shared.ts`.